src/stream/buffer_unordered.rs

```diff
@@ -39,11 +46,12 @@ impl<S> Stream for BufferUnordered<S>
 
     fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
         // First, try to fill in all the futures
-        for future in &mut self.futures {
+        for (idx, future) in self.futures.iter_mut().enumerate() {
```

Comment:
I wonder if this is perhaps why the perf isn't scaling? If this is full of idle connections we'll always fly over this array looking for more but we won't find any. Perhaps this could be converted to a `Slab`-style where we can check if a slot is open in O(1) time?